File: src/main/java/com/vld0/spark/PageSize.java
package com.vld0.spark;

import org.apache.commons.io.IOUtils;
import org.apache.http.HttpEntity;
import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClients;
import org.apache.spark.api.java.function.MapFunction;

import java.io.IOException;

public class PageSize implements MapFunction<String, Long> {

    @Override
    public Long call(String url) {
        CloseableHttpClient httpclient = HttpClients.createDefault();
        HttpGet get = new HttpGet(url);
        try (CloseableHttpResponse response = httpclient.execute(get)) {
            HttpEntity entity = response.getEntity();
            if (entity != null) {
                byte[] content = IOUtils.toByteArray(entity.getContent());
                //System.out.println(Strings.getString(content, Charset.defaultCharset().toString()));
                return (long)content.length;
            }
        } catch (IOException e) {
            return -1l;
        }

        return -1l;
    }

}

File: src/main/scala/vld0/PageSize.scala
package vld0

import java.io.IOException

import com.creanga.sparktest.LazyLogging
import org.apache.commons.httpclient.methods.GetMethod
import org.apache.commons.httpclient.{HttpClient, MultiThreadedHttpConnectionManager}
import org.apache.spark.{SparkEnv, TaskContext}

object PageSize extends LazyLogging {

  val connectionManager = new MultiThreadedHttpConnectionManager();
  val httpClient = new HttpClient(connectionManager)

  def call(url: String): Int = {

    val get = new GetMethod(url)
    get.setFollowRedirects(true)
    try {

      println(s"Processing url ${url} on ${SparkEnv.get.executorId} for partition ${TaskContext.getPartitionId()}")

      //System.out.println("connectionManager="+connectionManager.toString);
      //System.out.println("httpClient="+httpClient);

      //System.out.println("processing url="+url);

      logger.info("Processing url {} on {} for partition {}", url, SparkEnv.get.executorId, TaskContext.getPartitionId());


      httpClient.executeMethod(get)
      val content = get.getResponseBody
      return content.length

    } catch {
      case e: IOException =>
        return -1
    } finally {
      if (get!=null) {
        get.releaseConnection()
      }
    }
  }


}
